File: app/src/main/java/io/github/zwieback/familyfinance/business/chart/dialog/HorizontalBarChartDisplayDialog.java
package io.github.zwieback.familyfinance.business.chart.dialog;

import android.os.Bundle;
import android.support.annotation.NonNull;

import io.github.zwieback.familyfinance.R;
import io.github.zwieback.familyfinance.business.chart.display.HorizontalBarChartDisplay;
import io.github.zwieback.familyfinance.business.chart.display.type.HorizontalBarChartGroupByType;
import io.github.zwieback.familyfinance.business.chart.exception.UnsupportedHorizontalBarChartGroupByTypeException;
import io.github.zwieback.familyfinance.databinding.DialogDisplayChartBarHorizontalBinding;

import static io.github.zwieback.familyfinance.business.chart.display.HorizontalBarChartDisplay.HORIZONTAL_BAR_CHART_DISPLAY;

public class HorizontalBarChartDisplayDialog extends ChartDisplayDialog<HorizontalBarChartDisplay,
        DialogDisplayChartBarHorizontalBinding> {

    public static HorizontalBarChartDisplayDialog newInstance(HorizontalBarChartDisplay display) {
        HorizontalBarChartDisplayDialog fragment = new HorizontalBarChartDisplayDialog();
        Bundle args = createArguments(HORIZONTAL_BAR_CHART_DISPLAY, display);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    protected HorizontalBarChartDisplay createCopyOfDisplay(HorizontalBarChartDisplay display) {
        return new HorizontalBarChartDisplay(display);
    }

    protected String getInputDisplayName() {
        return HORIZONTAL_BAR_CHART_DISPLAY;
    }

    @Override
    protected int getDialogTitle() {
        return R.string.horizontal_bar_chart_display_title;
    }

    @Override
    protected int getDialogLayoutId() {
        return R.layout.dialog_display_chart_bar_horizontal;
    }

    protected void bind(HorizontalBarChartDisplay display) {
        binding.setDisplay(display);
    }

    protected void updateDisplayProperties() {
        display.setGroupByType(determineGroupByType());
    }

    @NonNull
    private HorizontalBarChartGroupByType determineGroupByType() {
        if (binding.groupByDays.isChecked()) {
            return HorizontalBarChartGroupByType.ARTICLE;
        } else if (binding.groupByWeeks.isChecked()) {
            return HorizontalBarChartGroupByType.ARTICLE_PARENT;
        }
        throw new UnsupportedHorizontalBarChartGroupByTypeException();
    }
}

File: app/src/main/java/io/github/zwieback/familyfinance/business/chart/dialog/HorizontalBarChartDisplayDialog.kt
package io.github.zwieback.familyfinance.business.chart.dialog

import androidx.annotation.StringRes
import io.github.zwieback.familyfinance.R
import io.github.zwieback.familyfinance.business.chart.display.HorizontalBarChartDisplay
import io.github.zwieback.familyfinance.business.chart.display.HorizontalBarChartDisplay.Companion.HORIZONTAL_BAR_CHART_DISPLAY
import io.github.zwieback.familyfinance.business.chart.display.type.HorizontalBarChartGroupByType
import io.github.zwieback.familyfinance.business.chart.exception.UnsupportedHorizontalBarChartGroupByTypeException
import io.github.zwieback.familyfinance.databinding.DialogDisplayChartBarHorizontalBinding

class HorizontalBarChartDisplayDialog :
    ChartDisplayDialog<HorizontalBarChartDisplay, DialogDisplayChartBarHorizontalBinding>() {

    override val inputDisplayName: String
        get() = HORIZONTAL_BAR_CHART_DISPLAY

    override val dialogTitle: Int
        get() = R.string.horizontal_bar_chart_of_expenses_display_title

    override val dialogLayoutId: Int
        get() = R.layout.dialog_display_chart_bar_horizontal

    override fun createCopyOfDisplay(display: HorizontalBarChartDisplay): HorizontalBarChartDisplay {
        return HorizontalBarChartDisplay(display)
    }

    override fun bind(display: HorizontalBarChartDisplay) {
        binding.display = display
    }

    override fun updateDisplayProperties() {
        display.groupByType = determineGroupByType()
    }

    private fun determineGroupByType(): HorizontalBarChartGroupByType {
        return when {
            binding.groupByDays.isChecked -> HorizontalBarChartGroupByType.ARTICLE
            binding.groupByWeeks.isChecked -> HorizontalBarChartGroupByType.ARTICLE_PARENT
            else -> throw UnsupportedHorizontalBarChartGroupByTypeException()
        }
    }

    companion object {
        fun newInstance(
            display: HorizontalBarChartDisplay,
            @StringRes dialogTitleId: Int
        ): HorizontalBarChartDisplayDialog {
            return HorizontalBarChartDisplayDialog().apply {
                arguments = createArguments(HORIZONTAL_BAR_CHART_DISPLAY, display, dialogTitleId)
            }
        }
    }
}
